File: Python3/283.Move Zeroes.py
"""
@Title: 283. Move Zeroes
@Tag: array
@Date: Jan-01 2020
@Author: ceezyyy
@Difficulty: Easy
"""


class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # corner case
        if len(nums) == 0 or len(nums) == 1:
            return
        index = 0  # record the number of the non-zero elements
        for i in range(len(nums)):
            if nums[i] != 0:
                # maintaining the relative order of the non-zero elements
                nums[index] = nums[i]
                index += 1  # count
        while index < len(nums):  # move all 0's to the end
            nums[index] = 0
            index += 1


"""
Runtime: 48 ms, faster than 86.96% of Python3 online submissions for Move Zeroes.
Memory Usage: 13.8 MB, less than 100.00% of Python3 online submissions for Move Zeroes.
"""


"""
Time Complexity: O(n)
Space Complexity: O(1)
"""

File: Others/cpp/283.Move Zeroes.cpp
/**
 * Source: LeetCode 283. Move Zeroes
 * Author: ceezyyy
 * Date: 03-26-2019
 */


/*
Easy

Share
Given an array nums, write a function to move all 0¡¯s to the end of it 
while maintaining the relative order of the non-zero elements.

Example:
Input: [0,1,0,3,12]
Output: [1,3,12,0,0]

Note:
You must do this in-place without making a copy of the array.
Minimize the total number of operations.

*/


#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
	void moveZeroes(vector<int>& nums) {
		int res = 0; //The slow-runner
		for (int i = 0; i < nums.size(); i++) { //The fast-runner
			if (nums.at(i) != 0) {
				nums.at(res++) = nums.at(i);
			}
		}
		while (res < nums.size()) {
			nums.at(res++) = 0;
		}
	}
};


// Study Notes:
// one slow-runner and one fast-runner at the same time


/**
 * Submission Detail:
 * Runtime: 16 ms, faster than 99.47% of C++ online submissions for Move Zeroes.
 * Memory Usage: 9.4 MB, less than 98.83% of C++ online submissions for Move Zeroes.
 */
